File: 12455.py

# 0 ≤ t ≤ 50，表示測資的筆數
t = int(input())
for iter1 in range(t):
    # 每筆測資三行，第一行有一個數字 n, 0 ≤ n ≤ 1000，表示我們所要的長度。
    n = int(input())
    # 第二行有一個數字 p, 1 ≤ p ≤ 20，表示我們所擁有的金屬棒的數量。
    p = int(input())
    # 第三行有 p 個數字，表示 p 根金屬棒的長度。
    P = list(map(int, input().split()))
 
    dp = [0 for i in range(n+1)]
    for i in P:
        for j in range(n, 0, -1):
            if (j >= i):
                dp[j] = max(dp[j], dp[j - i] + i)      
        print(dp)
 
    if (dp[n] == n):
        print("YES")
    else:
        print("NO")

File: 12455.cpp
/* a522: 12455 - Bars */
#include <iostream>
#include <algorithm>
using namespace std;

int main(){
    int T;
    cin>>T;
    for(int t=0;t<T;t++){
        int n,p;
        cin>>n;
        cin>>p;
        int A[p], dp[n+1]={};
        for(int i=0;i<p;i++) cin>>A[i];
        
        for(int i=0;i<p;i++){
            for(int j=n;j>-1;j--){
                if(j>=A[i]){
                    dp[j]=max(dp[j],dp[j-A[i]]+A[i]);
                }
            }
        }
        if(dp[n]==n) cout<<"YES"<<endl;
        else cout<<"NO"<<endl;
    }

    return 0;
}